### PR TITLE
new structure with smi and install commands

### PR DESCRIPTION
**Pull request by kumarabd**
Signed-off-by: kumarabd <abishekkumar92@gmail.com>

### CODE REVIEW

**Comment by leecalcote** [comment]
Conflicts - https://github.com/layer5io/meshery-kuma/pull/71/conflicts

**Comment by kumarabd** [comment]
@mgfeller @kushthedude  It would be great if we can discuss on the new structure and the principles that needs to be defined to follow them throughout the adaptor repositories.

File: scripts/smi/smi.yaml [hunk 1]
```diff
@@ -0,0 +1,86 @@
+---
```
Comment (leecalcote) [inline scripts/smi/smi.yaml:1]: Can this be centrally defined and imported from the Learn Layer5 repo?

**Comment by leecalcote** [comment]
Thank you for today’s demo, @kumarabd. Before merging, will you attend to each of the failed build checks?

File: scripts/smi/delete.sh [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#!/bin/sh
+
+set -e
+
+if ! type "kubectl" > /dev/null 2>&1; then
+  printf "ERROR\tgrep cannot be found\n"
```
Comment (mgfeller) [inline scripts/smi/delete.sh:6]: kubectl cannot be found

File: scripts/smi/smi.yaml [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+          protocol: TCP
+        resources:
+          limits:
+            cpu: "500m"
+            memory: 1000Mi
+          requests:
+            cpu: "200m"
+            memory: 500Mi
+      restartPolicy: Always
+      serviceAccount: meshery
```
Comment (mgfeller) [inline scripts/smi/smi.yaml:55]: serviceAccount is deprecated I just learned from my IDE

File: scripts/smi/deploy.sh [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#!/bin/sh
+
+set -e
+
+if ! type "kubectl" > /dev/null 2>&1; then
+  printf "ERROR\tgrep cannot be found\n"
```
Comment (mgfeller) [inline scripts/smi/deploy.sh:6]: kubectl cannot be found

**Comment by mgfeller** [comment]
Hi @kumarabd, great demo today! I'm still familiarizing myself with the code, it looks good and I learn a lot :-) 

I also ran the adapter locally (microk8s).

File: scripts/smi/deploy.sh [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  exit 1;
+fi
+
+printf "INFO\tStarting SMI deployment......\n"
+if ! kubectl apply -f ./scripts/smi/smi.yaml; then
```
Comment (mgfeller) [inline scripts/smi/deploy.sh:11]: I wonder whether we have some guidelines when to use command line tool, and when to use client in the code, if one exists, like for kubectl? The consul adapter is using the kubectl client. Advantages and disadvantages of each approach, is e.g. error handling and error message propagation easier with a client?

File: internal/config/defaults.go [hunk 6]
```diff
@@ -24,24 +23,31 @@ var (
 
 	// operations holds the supported operations inside mesh
 	operations = map[string]interface{}{
-		installKuma: map[string]interface{}{
-			"type": "INSTALL",
+		InstallKuma: map[string]interface{}{
```
Comment (leecalcote) [inline internal/config/defaults.go:26]: Do adapter operations belong in a `defaults` file?

File: scripts/smi/deploy.sh (hunk 5 above)
Comment (leecalcote) [reply scripts/smi/deploy.sh:11]: @mgfeller I had the same concern. My default perspective is to always use the client built for the language (use Kubernetes client go). 

The crux of the consideration in this circumstance is the lack of a kuma client go, I believe, and the lack of public Kubernetes manifests? I think that we should investigate two things:

1. Availability of Kubernetes manifests for Kuma - and consider using kubernetes go client with these manifests
1. If for whatever reason the Kuma project is not publishing their Kubernetes deployment manifests (which would be _really_ odd), we could still use Kubernetes client go after piping the output of this command to a file / object: `kumactl install control-plane | kubectl apply -f -`

Comment (mgfeller) [reply scripts/smi/deploy.sh:11]: @lee ok, good to know, this default perspective makes sense. 
Kuma can be installed using Helm too https://kuma.io/docs/0.7.1/installation/helm/.

If we want to Helm functionality (and other stuff) common to all adapters (as discussed elsewhere), where would that code live? Learn Layer 5 repo like the smi.yaml as mentioned above? Or a new repo? Sorry if I have overlooked that in a doc or on slack :thinking:

Comment (kumarabd) [reply scripts/smi/deploy.sh:11]: @mgfeller  Totally agree with the above. I am currently pushing changes for client-go and helm client. Also about the helm chart, we are moving towards supporting multiple ways to deploy the mesh. The user decides which way he wants the mesh to be installed.

File: internal/config/defaults.go (hunk 6 above)
Comment (kumarabd) [reply internal/config/defaults.go:26]: @leecalcote Its either inject the config, or have it remained as a default. Injecting would mean that u do it via the `main.go` file. Do we want to have another hop for this?

File: scripts/smi/deploy.sh (hunk 5 above)
Comment (mgfeller) [reply scripts/smi/deploy.sh:11]: @kumarabd :+1: looking forward to seeing your code using the helm client, it's on the list for the consul adapter, hopefully able to work on that soon.

Comment (kumarabd) [reply scripts/smi/deploy.sh]: @mgfeller  The helm changes are pushed.

Comment (mgfeller) [reply scripts/smi/deploy.sh]: @kumarabd :+1:

Comment (leecalcote) [reply scripts/smi/deploy.sh]: Adding the default perspectives above as guiding principles in the Meshery Adapters design spec.

File: internal/config/defaults.go (hunk 6 above)
Comment (leecalcote) [reply internal/config/defaults.go]: Renaming this file to `supported_operations.go` will probably suffice.

Comment (mgfeller) [reply internal/config/defaults.go]: out of curiosity, do we have unsupported operations?

**Comment by leecalcote** [comment]
@kumarabd a few conflicts to resolve here - https://github.com/layer5io/meshery-kuma/pull/71/conflicts

**Comment by leecalcote** [comment]
@kumarabd good to merge after resolving conflicts. Let's release v0.2.0 after merge.

**Comment by kumarabd** [comment]
> @kumarabd good to merge after resolving conflicts. Let's release v0.2.0 after merge.

@leecalcote  Lets move with this first - https://github.com/layer5io/meshery-kuma/pull/75 so that @shivaylamba can move ahead with his task. We can merge this PR along with the our meshkit decision. Sounds good?

**Comment by leecalcote** [comment]
Sounds real good, @kumarabd 👍

**Pull request closed without merging**